docs/programming-oak.md

```diff
@@ -0,0 +1,305 @@
+  if (create_application_response == nullptr) {
+    LOG(QFATAL) << "Failed to create application";
+  }
+```
+
+The Oak Manager will launch an [Oak VM](concepts.md#oak-vm) inside the enclave,
+and this VM will check the provided Wasm module(s) and application
+configuration.  Assuming everything is correct (e.g. the Nodes all have an
+`oak_main` entrypoint and only expect to find the Oak [host
+functions](abi.md#host-functions)), the Oak VM opens up a port of its own and
+the `CreateApplication` returns this to the client.
+
+The client can now connect to this separate gRPC service, and send
+(Node-specific) gRPC requests to it, over a channel that has end-to-end
+encryption into the enclave:
+
+[embedmd]:# (../examples/rustfmt/client/rustfmt.cc C++ /.*InitializeAssertionAuthorities/ /CredentialsOption/)
+```C++
+  oak::ApplicationClient::InitializeAssertionAuthorities();
+
+  // Connect to the newly created Oak Application.
+  auto stub = FormatService::NewStub(grpc::CreateChannel(
+      addr.str(), asylo::EnclaveChannelCredentials(asylo::BidirectionalNullCredentialsOption
+```
+
+## Request Processing Path
+
+At this point, the client code can interact with the Node code via gRPC.  A
+typical sequence for this (using the various helpers described in previous
+sections) would be as follows:
+
+ - The Node code (Wasm code running in a Wasm interpreter, running in an
+   enclave) is blocked inside a call to the `oak.wait_on_channels()` host
+   function from the `oak::grpc::event_loop` helper function.
+    - `event_loop()` was invoked directly from the auto-generated `oak_main()`
+      exported function.
+ - The client C++ code builds a gRPC request and sends it to the Oak VM.
+    - This connection is end-to-end encrypted using the Asylo key exchange
+      mechanism.
+ - The Oak VM receives the message and encapsulates it in a `GrpcRequest`
+   wrapper message.
+ - The Oak VM serializes the `GrpcRequest` and writes it to the gRPC-in channel
+   for the node.
+ - This unblocks the Node code, and `oak::grpc::event_loop` reads and
+   deserializes the incoming gRPC request. It then calls the Node's `invoke()`
+   method with the method name and (serialized) gRPC request.
+ - The Node's `invoke()` method passes this straight on to the auto-generated
+   gRPC service code in `proto::service_name_grpc::dispatch()`.
+ - The auto-generated `dispatch()` code invokes the relevant method on the
+   `Node`.
+ - The (user-written) code in this method does its work, and returns a
+   response.
+ - The auto-generated `dispatch()` code encapsulates the response into a
+   `GrpcResponse` wrapper message, and serializes into the gRPC output channel.
+ - The Oak VM reads this message from the channel, deserializes it and sends the
+   inner response back to the client.
+ - The client C++ code receives the response.
+
+
+<!-- From (Google-internal): http://go/sequencediagram/view/5741464478810112 -->
```

Comment:
Google-internal?

Reply:
Yeah, non-Googlers can't follow that link but I think it's still helpful to indicate provenance where possible, so that people (well, Googlers) other than me can update the diagram in future.